File: lib/motion.ts
/**
 * Petsona — Motion tokens + reduce-motion handling.
 *
 * Five duration constants matching `tokens.css --motion-*`:
 *  - `instant`  60ms   — tap feedback, press scale
 *  - `fast`     150ms  — toggles, chips, color shifts
 *  - `medium`   260ms  — screen transitions, card reveals
 *  - `slow`     420ms  — splash → app, plan-day reveal
 *  - `languid`  700ms  — loading shimmers, decorative motion
 *
 * Two easing curves built from `Easing.bezier(...)` so consumers can pass
 * them straight to Reanimated's `withTiming`. Bezier values mirror
 * `tokens.css --easing-{primary,out}` exactly:
 *  - `primary` cubic-bezier(0.32, 0.72, 0.0, 1)
 *  - `out`     cubic-bezier(0.50, 0.00, 0.4, 1)
 *
 * `useReducedMotion()` returns the live value of
 * `AccessibilityInfo.isReduceMotionEnabled()`. Consumers branch on it to
 * drop durations to ~0ms when the user prefers reduced motion.
 *
 * Babel auto-config: `babel-preset-expo` auto-adds
 * `react-native-worklets/plugin` when the package is installed (verified
 * 2026-05-12 in node_modules/babel-preset-expo/build/index.js lines
 * 313-320). `babel.config.js` is NOT modified.
 */
import * as React from 'react';
import { AccessibilityInfo } from 'react-native';
import { Easing } from 'react-native-reanimated';

export const motion = {
  duration: {
    instant: 60,
    fast: 150,
    medium: 260,
    slow: 420,
    languid: 700,
  },
  easing: {
    primary: Easing.bezier(0.32, 0.72, 0.0, 1),
    out: Easing.bezier(0.5, 0.0, 0.4, 1),
  },
} as const;

export type MotionDurationKey = keyof typeof motion.duration;
export type MotionEasingKey = keyof typeof motion.easing;

/**
 * Subscribes to `AccessibilityInfo.reduceMotionChanged` and returns the live
 * state. Consumers should treat `true` as "do not animate" (set durations
 * to 0 or skip the animation entirely).
 */
export function useReducedMotion(): boolean {
  const [reduce, setReduce] = React.useState(false);
  React.useEffect(() => {
    let cancelled = false;
    AccessibilityInfo.isReduceMotionEnabled()
      .then((value) => {
        if (!cancelled) setReduce(value);
      })
      .catch(() => {
        // Platforms without the capability default to off.
      });
    const sub = AccessibilityInfo.addEventListener('reduceMotionChanged', setReduce);
    return () => {
      cancelled = true;
      sub.remove();
    };
  }, []);
  return reduce;
}

File: ios/Petsona/Petsona/DesignSystem/Motion.swift
import SwiftUI

public enum Motion {

    // MARK: - Durations (seconds, from tokens.css --motion-*)
    public static let instant: Double = 0.06
    public static let fast:    Double = 0.15
    public static let medium:  Double = 0.26
    public static let slow:    Double = 0.42
    public static let languid: Double = 0.70

    // MARK: - Easing (cubic-bezier from tokens.css --easing-*)
    // --easing-primary: cubic-bezier(0.32, 0.72, 0.0, 1)
    public static let primaryAnimation = Animation.timingCurve(0.32, 0.72, 0.0, 1, duration: medium)

    // --easing-out: cubic-bezier(0.50, 0.00, 0.4, 1)
    public static let outAnimation = Animation.timingCurve(0.50, 0.00, 0.4, 1, duration: medium)

    // MARK: - Reduce-motion-aware helper
    public static func animation(for duration: Double, reduced: Bool) -> Animation {
        reduced
            ? .linear(duration: 0.001)
            : .timingCurve(0.32, 0.72, 0.0, 1, duration: duration)
    }
}

// MARK: - View modifier for reduce-motion-aware animations

public extension View {
    func petsonaAnimation<V: Equatable>(
        _ value: V,
        duration: Double = Motion.medium
    ) -> some View {
        modifier(ReduceMotionAnimation(value: value, duration: duration))
    }
}

private struct ReduceMotionAnimation<V: Equatable>: ViewModifier {
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    let value: V
    let duration: Double

    func body(content: Content) -> some View {
        content.animation(
            Motion.animation(for: duration, reduced: reduceMotion),
            value: value
        )
    }
}
